pkg/k8s/pipelineRun.go

```diff
@@ -128,46 +128,55 @@ func (r *pipelineRun) UpdateState(state api.State) (*api.StateItem, error) {
 	r.ensureCopy()
 	klog.V(3).Infof("Update State to %s [%s]", state, r.String())
 	now := metav1.Now()
-	oldstate := r.finishCurrentState()
-	newState := api.StateItem{State: state, StartedAt: now}
-	if state == api.StateFinished {
-		newState.FinishedAt = now
-	}
-	return oldstate, r.changeStatusAndUpdateSafely(func() {
+	oldState := r.apiObj.Status.StateDetails
+
+	err := r.changeStatusAndUpdateSafely(func() error {
+
+		currentState := r.apiObj.Status.StateDetails
+		if currentState != oldState {
+			return fmt.Errorf("State cannot be updated as it was changed concurrently from %q to %q", oldState, currentState)
+		}
+		if currentState.State == api.StateUndefined {
+			currentState.State = api.StateNew
+			currentState.StartedAt = r.apiObj.ObjectMeta.CreationTimestamp
+			r.apiObj.Status.StartedAt = &now
+		}
+		currentState.FinishedAt = now
+		his := r.apiObj.Status.StateHistory
+		his = append(his, currentState)
+
+		newState := api.StateItem{State: state, StartedAt: now}
+		if state == api.StateFinished {
+			newState.FinishedAt = now
+		}
+
 		r.apiObj.Status.StateDetails = newState
+		r.apiObj.Status.StateHistory = his
 		r.apiObj.Status.State = state
+		return nil
 	})
+
+	if err != nil {
+		return nil, err
+	}
+	oldState.State = api.StateNew
+	oldState.StartedAt = r.apiObj.ObjectMeta.CreationTimestamp
+	return &oldState, nil
```

Comment:
This is wrong in case the initial state was set already.

Reply:
Take result from history now and improved tests.